handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -151,14 +151,24 @@
         SET_SERVER_NAMES_METHOD = setServerNamesMethod;
     }
 
-    private static final int MAX_PLAINTEXT_LENGTH = 16 * 1024; // 2^14
-    private static final int MAX_COMPRESSED_LENGTH = MAX_PLAINTEXT_LENGTH + 1024;
-    private static final int MAX_CIPHERTEXT_LENGTH = MAX_COMPRESSED_LENGTH + 1024;
+    static final int MAX_PLAINTEXT_LENGTH = 16 * 1024; // 2^14
 
-    // Header (5) + Data (2^14) + Compression (1024) + Encryption (1024) + MAC (20) + Padding (256)
-    static final int MAX_ENCRYPTED_PACKET_LENGTH = MAX_CIPHERTEXT_LENGTH + 5 + 20 + 256;
+    /**
+     * This is the maximum overhead when encrypting plaintext as defined by
+     * <a href="https://www.ietf.org/rfc/rfc5246.txt">rfc5264</a>,
+     * <a href="https://www.ietf.org/rfc/rfc5289.txt">rfc5289</a> and openssl implementation itself.
+     *
+     * Please note that we use a padding of 16 here as openssl uses PKC#5 which uses 16 bytes while the spec itself
+     * allow up to 255 bytes. 16 bytes is the max for PKC#5 (which handles it the same way as PKC#7) as we use a block
+     * size of 16. See <a href="https://tools.ietf.org/html/rfc5652#section-6.3">rfc5652#section-6.3</a>.
+     *
+     * 16 (IV) + 48 (MAC) + 1 (Padding_length field) + 16 (Padding) + 1 (ContentType) + 2 (ProtocolVersion) + 2 (Length)
```

Comment:
The goal when padding is to get up to a multiple of the block size (say 16 for AES). So if my input has length, say, 26, I want 6 total bytes of padding which is 5 byte of "padding" and one byte of "padding length".

If you only ever add the minimal padding needed, as OpenSSL does, you will add somewhere between 0 and 15 bytes of padding, or somewhere between 0 + 1 = 1 and 15 + 1 = 16 bytes total. You wouldn't add 16 + 1 = 17 bytes of padding because you instead would have added 0 + 1 = 1 bytes.